packages/math-input/src/components/__tests__/integration.test.tsx

```diff
@@ -0,0 +1,183 @@
+                            cb();
+                        }}
+                        onFocus={() => {
+                            keypadElement?.activate();
+                        }}
+                        onBlur={() => {
+                            keypadElement?.dismiss();
+                        }}
+                    />
+                );
+            }}
+        </KeypadContext.Consumer>
+    );
+}
+
+function KeypadWithContext() {
+    return (
+        <KeypadContext.Consumer>
+            {({setKeypadElement}) => {
+                return (
+                    <KeypadSwitch
+                        onElementMounted={setKeypadElement}
+                        onDismiss={() => {}}
+                        onAnalyticsEvent={async () => {}}
+                        useV2Keypad
+                    />
+                );
+            }}
+        </KeypadContext.Consumer>
+    );
+}
+
+function ConnectedMathInput() {
+    const [keypadElement, setKeypadElement] = useState<KeypadAPI | null>();
+    const [renderer, setRenderer] = useState<any>(null);
+    const [scrollableElement, setScrollableElement] =
+        useState<HTMLElement | null>();
+
+    return (
+        <KeypadContext.Provider
+            value={{
+                setKeypadElement,
+                keypadElement,
+                setRenderer,
+                renderer,
+                setScrollableElement,
+                scrollableElement,
+            }}
+        >
+            <InputWithContext />
+            <KeypadWithContext />
+        </KeypadContext.Provider>
+    );
+}
+
+describe("math input integration", () => {
+    it("renders", () => {
+        render(<ConnectedMathInput />);
+
+        expect(
+            screen.getByLabelText(
+                "Math input box Tap with one or two fingers to open keyboard",
+            ),
+        ).toBeInTheDocument();
+    });
+
+    it("doesn't show the keypad initially", () => {
+        render(<ConnectedMathInput />);
+
+        expect(
+            screen.queryByRole("button", {name: "1"}),
+        ).not.toBeInTheDocument();
+    });
+
+    it("shows the keypad after input interaction", async () => {
+        render(<ConnectedMathInput />);
+
+        const input = screen.getByLabelText(
+            "Math input box Tap with one or two fingers to open keyboard",
+        );
+
+        fireEvent.touchStart(input);
+
+        expect(screen.getByRole("button", {name: "1"})).toBeVisible();
+    });
+
+    it("updates input when using keypad", () => {
+        render(<ConnectedMathInput />);
+
+        const input = screen.getByLabelText(
+            "Math input box Tap with one or two fingers to open keyboard",
+        );
+
+        fireEvent.touchStart(input);
+        userEvent.click(screen.getByRole("button", {name: "1"}));
+
+        // MathQuill is problematic,
+        // this is the only way I know how to test the "input"
+        const mathquillInput =
```

Comment:
Makes sense! You might also be able to get a ref to the MathInput or check that setValue is called (since you're providing the method anyway) to get access to mathquill that way. Not necessary imo but might help clean things up.

Reply:
On the more complex tests I used `latex()` from MathQuill, but I kept the simple tests as-is. I think it's useful to display the complexity of MathQuill to highlight some of the problematic bits: what's shown to the user is a bunch of characters wrapped by spans, so I wanted some tests to represent that.